File: samples/Rask.Example.Shared/Features/Gantt/Gantt.cs
using System.Diagnostics.CodeAnalysis;
using System.Globalization;
using System.Text.Json;
using Microsoft.JSInterop;
using Rask.Core.Live;

namespace Rask.Example.Shared.Features;

/// <summary>One bar on the chart.</summary>
/// <param name="Id">Stable identity — what the event callbacks report back.</param>
/// <param name="Name">Label drawn on the bar.</param>
/// <param name="Start">First day of the task (inclusive).</param>
/// <param name="End">Last day of the task (inclusive).</param>
/// <param name="Progress">Completion, 0–100.</param>
public sealed record GanttTask(string Id, string Name, DateOnly Start, DateOnly End, double Progress);

/// <summary>A single non-working day, highlighted on the chart.</summary>
/// <param name="Date">The day to highlight.</param>
/// <param name="Label">Shown when the day is hovered.</param>
/// <remarks>
///     A labelled single day, not a range — that is frappe-gantt's own holiday model, and promising a
///     range here would be promising something the library can't draw. Pass consecutive days for a
///     multi-day closure.
/// </remarks>
public sealed record GanttHoliday(DateOnly Date, string Label);

/// <summary>A bar was dragged or resized.</summary>
/// <param name="TaskId">Which task moved.</param>
/// <param name="Start">Its new first day.</param>
/// <param name="End">Its new last day.</param>
/// <remarks>
///     One parameter, not three, and that is load-bearing rather than taste: the generated factory only
///     wraps a callback for auto-re-render when it takes <b>at most one</b> argument. A three-arg delegate
///     is passed through raw, so the owning component would not re-render when it fires and the caller
///     would have to reach for <c>StateHasChanged()</c>. Bundling the arguments keeps the callback on the
///     framework's normal path — which is why nothing in <see cref="GanttDemo" /> calls it.
/// </remarks>
public sealed record GanttDateChange(string TaskId, DateOnly Start, DateOnly End);

/// <summary>A bar's progress handle moved.</summary>
/// <param name="TaskId">Which task changed.</param>
/// <param name="Progress">Its new completion, 0–100.</param>
/// <remarks>One parameter for the same reason as <see cref="GanttDateChange" />.</remarks>
public sealed record GanttProgressChange(string TaskId, double Progress);

/// <summary>Chart zoom level. Maps to frappe-gantt's <c>view_mode</c>.</summary>
public enum GanttViewMode
{
    /// <summary>One column per hour.</summary>
    Hour,

    /// <summary>One column per six hours.</summary>
    QuarterDay,

    /// <summary>One column per twelve hours.</summary>
    HalfDay,

    /// <summary>One column per day. The library's default.</summary>
    Day,

    /// <summary>One column per week.</summary>
    Week,

    /// <summary>One column per month.</summary>
    Month,

    /// <summary>One column per year.</summary>
    Year
}

/// <summary>
///     A Gantt chart, wrapping the third-party <see href="https://github.com/frappe/gantt">frappe-gantt</see>
///     library (MIT, vendored under <c>wwwroot/lib/frappe-gantt</c>). Renders on both transports and on the
///     Native hosts, because everything host-specific is funnelled through <see cref="IJSRuntime" /> and
///     <see cref="RaskLiveOptions.PathBase" />.
/// </summary>
/// <remarks>
///     <para>
///         <b>This is showcase code, not a framework component</b> — it lives in the samples deliberately.
///         It's the worked example for the "third-party libraries" section of <c>docs/js-interop.md</c>:
///         copy it as the recipe for wrapping any DOM-owning JS library in Rask. It is not in
///         <c>Rask.Bootstrap</c> because that package ships zero JavaScript by design.
///     </para>
///     <para>
///         <b>The invariant that makes this safe:</b> the library owns every node inside the host div, and
///         the .NET side renders that div as a childless leaf. The normal diff addresses nodes by
///         positional path from the render tree, so it can never reach inside a leaf. Full-HTML frames
///         (the first interactive frame after page load always is one) are applied by morphing the
///         document, and a morph *would* pair the host's live children against the rendered zero and
///         delete the chart — so <c>Gantt.js</c> tags them <c>data-rask-managed</c>, which the reconciler
///         skips. Keep the host a leaf and keep that tag; drop either and the chart disappears.
///     </para>
/// </remarks>
public sealed partial class Gantt : Component
{
    private readonly ElementRef _host = ElementRef.New();
    private readonly IJSRuntime _js;
    private string? _id;
    private bool _mounted;

    // GanttInterop's [JSInvokable]s are reached only through the JS DotNet dispatcher (reflection), so
    // without this the WASM trimmer would drop them and every chart event would silently do nothing.
    [DynamicDependency(DynamicallyAccessedMemberTypes.PublicMethods, typeof(GanttInterop))]
    public Gantt(IJSRuntime js) => _js = js;

    /// <summary>The bars to draw. Required.</summary>
    public new required IReadOnlyList<GanttTask> Data { get; set; }

    /// <summary>Non-working days to highlight. Weekends are highlighted by the library regardless.</summary>
    public IReadOnlyList<GanttHoliday>? Holidays { get; set; }

    /// <summary>Zoom level. Changing it re-renders the chart in place, keeping scroll position.</summary>
    public GanttViewMode ViewMode { get; set; } = GanttViewMode.Day;

    /// <summary>Raised with the task id when a bar is clicked.</summary>
    public Func<string, Task>? OnTaskClick { get; set; }

    /// <summary>Raised when a bar is dragged or resized.</summary>
    public Func<GanttDateChange, Task>? OnDateChange { get; set; }

    /// <summary>Raised when a bar's progress handle moves.</summary>
    public Func<GanttProgressChange, Task>? OnProgressChange { get; set; }

    // A leaf: no children here, ever. See the class remarks — this is what keeps the diff out of the
    // library's DOM. The visible height comes from Gantt.css.
    protected override Component? Render() => Div.Ref(_host).Class("rask-gantt");

    // Mount in OnRendered, not OnMount: OnMount runs *before* the first render, so the host element
    // doesn't exist yet and the ref would resolve to null on the JS side. Interop issued during a render
    // walk is queued onto the frame and runs after the diff is applied, so the div is committed by then.
    protected override async Task OnRenderedAsync(bool firstRender)
    {
        if (!firstRender)
        {
            return;
        }

        _id = GanttInterop.Register(this);
        var mountedWith = BuildOptionsJson();
        try
        {
            await _js.InvokeVoidAsync("Rask.Gantt.mount", _host, _id, LiveOptions.PathBase, mountedWith);
        }
        catch
        {
            GanttInterop.Unregister(_id);
            _id = null;
            throw;
        }

        _mounted = true;

        // That await spans the library's own download, and a prop change landing inside that window found
        // _mounted still false and skipped its update. Reconcile once here or the chart is left showing
        // the state it mounted with, permanently, until some later unrelated change happens to push.
        var latest = BuildOptionsJson();
        if (!string.Equals(latest, mountedWith, StringComparison.Ordinal))
        {
            await _js.InvokeVoidAsync("Rask.Gantt.update", _host, latest);
        }
    }

    // Fires only on a real prop change, so this is the natural place to push new data at the library
    // rather than re-mounting it (which would lose scroll position).
    protected override async Task OnPropsChangedAsync()
    {
        if (_mounted)
        {
            await _js.InvokeVoidAsync("Rask.Gantt.update", _host, BuildOptionsJson());
        }
    }

    // Teardown is sync and fire-and-forget on purpose. An IAsyncDisposable component is *awaited* by the
    // framework's dispose walk, and that walk also runs when a session is torn down because its socket
    // already closed — at which point an interop call has nobody to answer it and never completes. There
    // is no ambient timeout, so awaiting here would hang the session's disposal forever and leak its DI
    // scope, its whole component tree and its locks: once per visitor, until the process runs out.
    protected override void OnUnmount()
    {
        if (_id is { } id)
        {
            GanttInterop.Unregister(id);
            _id = null;
        }

        if (!_mounted)
        {
            return;
        }

        _mounted = false;
        _ = DestroyQuietlyAsync();
    }

    private async Task DestroyQuietlyAsync()
    {
        try
        {
            await _js.InvokeVoidAsync("Rask.Gantt.destroy", _host);
        }
        catch
        {
            // The client is already gone (tab closed, socket dropped) — its DOM went with it, so there is
            // nothing left to tear down and nothing useful to report. Never let this reach the void.
        }
    }

    internal Task HandleTaskClickAsync(string taskId) => OnTaskClick?.Invoke(taskId) ?? Task.CompletedTask;

    internal Task HandleDateChangeAsync(string taskId, string startIso, string endIso)
    {
        if (OnDateChange is null || !TryParseIsoDate(startIso, out var start) || !TryParseIsoDate(endIso, out var end))
        {
            return Task.CompletedTask;
        }

        return OnDateChange(new GanttDateChange(taskId, start, end));
    }

    internal Task HandleProgressChangeAsync(string taskId, double progress) =>
        OnProgressChange?.Invoke(new GanttProgressChange(taskId, Math.Clamp(progress, 0, 100)))
        ?? Task.CompletedTask;

    // frappe-gantt's view_mode values are its own display strings — two of them have a space, which no
    // enum-name convention would produce. Map explicitly rather than deriving from the name.
    internal static string ToJsViewMode(GanttViewMode mode) => mode switch
    {
        GanttViewMode.Hour => "Hour",
        GanttViewMode.QuarterDay => "Quarter Day",
        GanttViewMode.HalfDay => "Half Day",
        GanttViewMode.Day => "Day",
        GanttViewMode.Week => "Week",
        GanttViewMode.Month => "Month",
        GanttViewMode.Year => "Year",
        _ => "Day"
    };

    // Serialize to a JSON string and hand *that* across, rather than letting the interop layer marshal a
    // complex type — the WASM showcase publishes trimmed, and this keeps the boundary to strings only.
    internal string BuildOptionsJson()
    {
        var payload = new GanttJsOptions(
            ToJsViewMode(ViewMode),
            [.. Data.Select(t => new GanttJsTask(t.Id, t.Name, ToIso(t.Start), ToIso(t.End), t.Progress))],
            [.. (Holidays ?? []).Select(h => new GanttJsHoliday(ToIso(h.Date), h.Label))]);

        return JsonSerializer.Serialize(payload, GanttJsonContext.Default.GanttJsOptions);
    }

    // The WASM showcase runs InvariantGlobalization, so pin the culture rather than inheriting the
    // ambient one — an ISO date is the contract with the library either way.
    private static string ToIso(DateOnly date) => date.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);

    private static bool TryParseIsoDate(string value, out DateOnly date)
    {
        // Gantt.js sends an unzoned wall-clock timestamp and we only model whole days, so keep the date
        // part as written. RoundtripKind (not None) is what stops a value that *does* carry a zone from
        // being converted to this machine's local time, which would land the bar on the wrong day — the
        // server's timezone has no business deciding which column the user dropped a bar on.
        if (DateTime.TryParse(value, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out var parsed))
        {
            date = DateOnly.FromDateTime(parsed);
            return true;
        }

        date = default;
        return false;
    }
}

File: samples/Rask.Example.Shared/Features/Gantt/Gantt.ts
// Scoped TypeScript for the Gantt component — the browser half of wrapping a third-party DOM library.
//   * mount(host, id, pathBase, optionsJson) — load frappe-gantt, draw the chart into the host, and
//     forward its callbacks to the [JSInvokable]s in GanttInterop.
//   * update(host, optionsJson) — push new tasks / a new view mode at the live instance.
//   * destroy(host) — drop the instance and clear the host.
//
// frappe-gantt is vendored under wwwroot/lib/frappe-gantt (MIT). It's an ordinary UMD bundle: loading it
// sets a `Gantt` global, described in the sibling frappe-gantt.d.ts because there is no node_modules to
// take typings from. Vendored rather than CDN-loaded so the showcase works offline and under the
// GitHub Pages sub-path.

const ASSEMBLY = "Rask.Example.Shared";

/** What the .NET side sends over as `optionsJson`. */
interface GanttComponentOptions {
    viewMode: string;
    tasks: GanttTask[];
    holidays: GanttHoliday[];
}

/** One live chart, plus enough of its current state to tell a real change from a no-op. */
interface ChartEntry {
    chart: GanttChart;
    viewMode: string;
    tasksJson: string;
}

// One instance per host element. A WeakMap, so a host that goes away with the page doesn't pin its chart.
const charts = new WeakMap<HTMLElement, ChartEntry>();
let loadPromise: Promise<typeof GanttChart> | null = null;

// Resolve the library's URL from the PathBase the .NET side passes in, rather than from document.baseURI:
// PathBase is what's correct behind a reverse proxy (Server) and under /rask/ (WASM on Pages).
function libUrl(pathBase: string | null, file: string): string {
    return `${pathBase || ""}/_content/Rask.Example.Shared/lib/frappe-gantt/${file}`;
}

function loadGantt(pathBase: string | null): Promise<typeof GanttChart> {
    if (globalThis.Gantt) return Promise.resolve(globalThis.Gantt);
    if (loadPromise) return loadPromise;

    loadPromise = new Promise<typeof GanttChart>((resolve, reject) => {
        // The stylesheet goes into <head> at runtime, so it isn't part of any .NET render. Rask notices
        // foreign head nodes and preserves them across re-renders on its own — that's the behaviour this
        // very demo sits under in docs/js-interop.md, and why there's no marking code here.
        const css = document.createElement("link");
        css.rel = "stylesheet";
        css.href = libUrl(pathBase, "frappe-gantt.css");
        document.head.appendChild(css);

        const script = document.createElement("script");
        script.src = libUrl(pathBase, "frappe-gantt.umd.js");
        script.onload = () => {
            // The bundle sets the global as a side effect of running. If it somehow did not, reject
            // rather than resolving undefined — otherwise the failure surfaces later as "Lib is not a
            // constructor", pointing at the call site instead of at the load.
            const lib = globalThis.Gantt;
            if (lib) {
                resolve(lib);
            } else {
                loadPromise = null;
                reject(new Error("frappe-gantt loaded but set no global"));
            }
        };
        script.onerror = () => {
            // Drop the failed attempt and take the half-loaded tags with it. Memoizing a rejected promise
            // would turn one flaky fetch into a permanently broken chart for the rest of the page's life,
            // with no way back short of a reload.
            loadPromise = null;
            css.remove();
            script.remove();
            reject(new Error("frappe-gantt failed to load"));
        };
        document.head.appendChild(script);
    });
    return loadPromise;
}

// The chart's DOM is created by the library, so it is absent from every .NET render payload. Not every
// frame is a diff — the first interactive frame after page load ships full HTML, and the client applies
// a full frame by morphing the document. The morph pairs the host's live children against the rendered
// zero and deletes the chart. Remove this and the chart vanishes a moment after it first draws.
// The marker is the opt-out: the reconciler skips marked from-side nodes, so the host pairs as empty.
// Mark the library's children, never the host itself — the host IS in the render tree, and marking it
// would make the morph treat it as missing and append a duplicate.
function markManaged(host: HTMLElement): void {
    for (const child of Array.from(host.children)) child.setAttribute("data-rask-managed", "");
}

function toLibTasks(options: GanttComponentOptions): GanttTask[] {
    return options.tasks.map((t) => ({
        id: t.id,
        name: t.name,
        start: t.start,
        end: t.end,
        progress: t.progress
    }));
}

// frappe-gantt keys holidays by the CSS colour to paint them with; weekends are its default entry, so
// keep that and add ours alongside rather than replacing it.
function toLibHolidays(options: GanttComponentOptions): GanttHolidays {
    const holidays: GanttHolidays = { "var(--g-weekend-highlight-color)": "weekend" };
    if (options.holidays.length > 0) {
        holidays["#ffd7d7"] = options.holidays.map((h) => ({ date: h.date, label: h.label }));
    }
    return holidays;
}

export async function mount(
    host: HTMLElement | null,
    id: string,
    pathBase: string | null,
    optionsJson: string): Promise<void> {
    if (!host) return;
    // A host that already carries a chart means a *new* component instance landed on the same element
    // (the .NET side rebuilt it — e.g. its position among its parent's children shifted). Rebuild rather
    // than bailing out: returning early would leave the previous instance's chart on screen, wired to a
    // component that no longer renders, so it would ignore every later prop change.
    if (charts.has(host)) destroy(host);

    const options = JSON.parse(optionsJson) as GanttComponentOptions;
    let Lib: typeof GanttChart;
    try {
        Lib = await loadGantt(pathBase);
    } catch {
        // Say so rather than leaving an empty box. It goes in as a marked *element*, not a bare text
        // node: the .NET side renders this host empty, so anything unmarked in here is something the next
        // full-frame morph deletes — and markManaged only reaches elements.
        const note = document.createElement("p");
        note.className = "text-danger small mb-0";
        note.textContent = "The Gantt library could not be loaded.";
        host.replaceChildren(note);
        markManaged(host);
        return;
    }

    // Re-entrancy: mount is awaited, so a second call could have landed while the script was loading.
    if (charts.has(host)) return;

    const chart = new Lib(host, toLibTasks(options), {
        view_mode: options.viewMode,
        holidays: toLibHolidays(options),
        infinite_padding: false,
        popup_on: "click",
        on_click: (task) => {
            window.DotNet.invokeMethodAsync(ASSEMBLY, "RaskGanttTaskClicked", id, String(task.id));
        },
        on_date_change: (task, start, end) => {
            window.DotNet.invokeMethodAsync(
                ASSEMBLY, "RaskGanttDateChanged", id, String(task.id), localIso(start), localIso(end));
        },
        on_progress_change: (task, progress) => {
            window.DotNet.invokeMethodAsync(
                ASSEMBLY, "RaskGanttProgressChanged", id, String(task.id), Number(progress));
        }
    });

    // Remember what the chart is currently showing, so update() can apply only real changes.
    charts.set(host, { chart, viewMode: options.viewMode, tasksJson: JSON.stringify(toLibTasks(options)) });
    markManaged(host);
}

export function update(host: HTMLElement | null, optionsJson: string): void {
    const entry = host ? charts.get(host) : undefined;
    if (!entry || !host) return;

    const options = JSON.parse(optionsJson) as GanttComponentOptions;
    const tasks = toLibTasks(options);
    const tasksJson = JSON.stringify(tasks);

    // Apply only what actually changed. The library's refresh() already re-applies the current view mode
    // internally, so calling both unconditionally rebuilt the entire SVG twice per update — and did it on
    // every prop change, including ones that touched neither.
    if (entry.viewMode !== options.viewMode) {
        // maintain_pos: keep the user where they were scrolled to while the axis rescales.
        entry.chart.change_view_mode(options.viewMode, true);
        entry.viewMode = options.viewMode;
    }

    if (entry.tasksJson !== tasksJson) {
        entry.chart.refresh(tasks);
        entry.tasksJson = tasksJson;
    }

    // Rebuilt nodes are new nodes, so they need the marker again.
    markManaged(host);
}

export function destroy(host: HTMLElement | null): void {
    if (!host || !charts.has(host)) return;
    // The library has no destroy() of its own, so drop our reference and clear what it drew. The host
    // itself belongs to the .NET render tree — leave it in place.
    charts.delete(host);
    host.replaceChildren();
}

// The library hands back Date objects holding a *wall-clock* day — the bar sits on the column the user
// sees. toISOString() would convert that to UTC and shift the calendar day for every user who isn't on
// UTC (a bar dropped on the 8th coming back as the 9th, or the 7th, depending on which side you're on),
// so format the local fields instead and let the .NET side read them as an unzoned timestamp.
function localIso(date: Date): string {
    if (!(date instanceof Date)) return String(date);
    const pad = (n: number): string => String(n).padStart(2, "0");
    return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`
        + `T${pad(date.getHours())}:${pad(date.getMinutes())}:${pad(date.getSeconds())}`;
}
